pkg/morph/deploy/notary.go

```diff
@@ -0,0 +1,880 @@
+							zap.Error(err))
+						continue
+					}
+				} else {
+					tick, err = initDesignateNotaryRoleAsSignerTick(prm, monitor)
+					if err != nil {
+						prm.logger.Error("failed to construct action designating Notary role to the multi-acc committee as signer, will try again later",
+							zap.Error(err))
+						continue
+					}
+				}
+			}
+		}
+
+		tick()
+	}
+}
+
+// initDesignateNotaryRoleToLocalAccountTick returns a function that preserves
+// context of the Notary role designation to the local account between calls.
+func initDesignateNotaryRoleToLocalAccountTick(prm enableNotaryPrm, monitor *blockchainMonitor) (func(), error) {
+	_actor, err := actor.NewSimple(prm.blockchain, prm.localAcc)
+	if err != nil {
+		return nil, fmt.Errorf("init transaction sender from local account: %w", err)
+	}
+
+	roleContract := rolemgmt.New(_actor)
+
+	// multi-tick context
+	var sentTxValidUntilBlock uint32
+
+	return func() {
+		if sentTxValidUntilBlock > 0 && sentTxValidUntilBlock <= monitor.currentHeight() {
+			prm.logger.Info("previously sent transaction designating Notary role to the local account may still be relevant, will wait for the outcome")
+			return
+		}
+
+		if sentTxValidUntilBlock > 0 {
+			prm.logger.Info("transaction designating Notary role to the local account was sent earlier, checking relevance...")
+
+			if cur := monitor.currentHeight(); cur <= sentTxValidUntilBlock {
+				prm.logger.Info("previously sent transaction designating Notary role to the local account may still be relevant, will wait for the outcome",
+					zap.Uint32("current height", cur), zap.Uint32("retry after height", sentTxValidUntilBlock))
+				return
+			}
+
+			prm.logger.Info("previously sent transaction designating Notary role to the local account expired without side-effect")
+		}
+
+		prm.logger.Info("sending new transaction designating Notary role to the local account...")
+
+		var err error
+
+		_, vub, err := roleContract.DesignateAsRole(noderoles.P2PNotary, keys.PublicKeys{prm.localAcc.PublicKey()})
+		if err != nil {
+			sentTxValidUntilBlock = 0
+			if isErrNotEnoughGAS(err) {
+				prm.logger.Info("not enough GAS to designate Notary role to the local account, will try again later")
+			} else {
+				prm.logger.Error("failed to send transaction designating Notary role to the local account, will try again later", zap.Error(err))
+			}
+			return
+		}
+
+		sentTxValidUntilBlock = vub
+
+		prm.logger.Info("transaction designating Notary role to the local account has been successfully sent, will wait for the outcome")
+	}, nil
+}
+
+// initDesignateNotaryRoleAsLeaderTick returns a function that preserves context
+// of the Notary role designation to the multi-acc committee between calls. The
+// operation is performed by the leading committee member which is assigned to
+// collect signatures for the corresponding transaction.
+func initDesignateNotaryRoleAsLeaderTick(prm enableNotaryPrm, monitor *blockchainMonitor) (func(), error) {
```

Comment:
i think what func would not pass `funlen` linter

Reply:
afaik we don't use `funlen` linter now. I don't think splitting into several functions will improve readability, at least for now

Reply:
> afaik we don't use funlen linter now

yes, i will create an issue for that. i would not find it if the question was about ~60-120 but it is already 400 lines long. also, i did not measure that and did not _want_ to find smth long; i just was reading and understood that the func was not going to end and i had lost the initial context of the code i was reading

Reply:
function length is an eternal topic of discussion. Same as functions which are created to just decrease length of some function and called once

> lost the initial context of the code i was reading
i believe, but log messages tell u what's goin on, dont think separate functions would ease readability: u then need to jump into them and read docs and go back

Reply:
https://github.com/nspcc-dev/neofs-node/issues/2407